osf/management/commands/reindex_with_current_metrics_mappings.py

```diff
@@ -0,0 +1,80 @@
+"""
+Reindex data to use current mapping for ES metrics classes
+"""
+import logging
+
+from django.core.management.base import BaseCommand
+from elasticsearch_dsl import connections
+from elasticsearch_metrics.registry import registry
+
+logger = logging.getLogger(__name__)
+
+
+def get_metric_class(index_name: str) -> type:
+    app_label, model_name = index_name.split('_')[:2]
+    return registry.all_metrics[app_label][model_name]
+
+
+def increment_index_versions(old_indices: list):
+    """
+    Increment versions numbers for new indices, these kind don't matter because they should always be aliased to
+    the original format of {app_label}_{cls.__name__.lower()}_{year}.
+
+    :param old_indices: indices to be updated
+    :return: indices names that are going to be reindexed into.
+    """
+    new_indices = []
+    for index in old_indices:
+        if '_v' in index and index[-1].isdigit():
+            name, version_num = index.split('_v')
+            new_index = f'{name}_v{int(version_num) + 1}'
+        else:
+            new_index = f'{index}_v2'
+        new_indices.append(new_index)
+
+    return new_indices
+
+
+def reindex_and_alias(old_indices: list):
+    """
+    To migrate data in ES with new mappings is a 4 step process:
+    1) Create an index with new mappings
+    2) Reindex data from old to new
+    3) Delete the old index
+    4) Alias the new index so it references the old.
+
+    :param old_indices: indices with data that has old mappings
+    :return: None
+    """
+    client = connections.get_connection()
+    new_indices = increment_index_versions(old_indices)
+
+    for old_index, new_index in zip(old_indices, new_indices):
+        metric_class = get_metric_class(old_index)
+        client.indices.create(new_index, body=metric_class._index.to_dict(), params={'wait_for_active_shards': 1})
+        body = {
+            'source': {
+                'index': old_index
+            },
+            'dest': {
+                'index': new_index
+            }
+        }
+        client.reindex(body, params={'wait_for_completion': 'true'})
+        client.indices.delete(old_index)
```

Comment:
It would be preferable if deletion happened after aliasing, but there are some issues with that on the first run -- the initial index name would conflict with the alias. It may be possible to avoid this if we alias the old to `_v1` on the first run, but I haven't confirmed this.

Reply:
`RequestError(400, 'invalid_alias_name_exception', 'Invalid alias name [osf_preprintdownload_2020], an index exists with the same name as the alias')` Apparently you must delete before aliasing.

Reply:
Is that after aliasing the initial index to `{name}_v1`?

Regardless, shouldn't affect subsequent runs